plugins/client.go

```diff
@@ -0,0 +1,217 @@
+		if tlsconfig != nil && tlsconfig.InsecureSkipVerify == false {
+			config, err = client.GenTLSConfig(tlsconfig.KeyFile, tlsconfig.CertFile, tlsconfig.CAFile)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse plugin tls config")
+			}
+		} else {
+			config = &tls.Config{InsecureSkipVerify: true}
+		}
+	}
+
+	// The timeout set 0
+	httpCli := newHTTPClient(time.Duration(0), u, config)
+
+	return &PluginClient{
+		address: addr,
+		baseURL: baseURL,
+		client:  httpCli,
+	}, nil
+}
+
+// generateBaseURL generates the base url
+func generateBaseURL(u *url.URL) string {
+	if u.Scheme == "https" {
+		return "https://" + u.Host
+	} else if u.Scheme == "unix" {
+		return "http://d"
+	}
+
+	return "http://" + u.Host
+}
+
+// newHTTPClient generates a http client
+func newHTTPClient(timeout time.Duration, u *url.URL, tlsconfig *tls.Config) *http.Client {
+	tr := &http.Transport{
+		TLSClientConfig: tlsconfig,
+	}
+
+	switch u.Scheme {
+	case "unix":
+		dialer := func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return net.DialTimeout("unix", u.Path, timeout)
+		}
+		tr.DialContext = dialer
+	default:
+		dialer := func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return net.DialTimeout(network, addr, timeout)
+		}
+		tr.DialContext = dialer
+	}
+
+	return &http.Client{
+		Transport: tr,
+		Timeout:   timeout,
+	}
+}
+
+// newPluginRequest generates an plugin request
+func (cli *PluginClient) newPluginRequest(path string, data io.Reader) (*http.Request, error) {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	fullPath := cli.baseURL + path
+
+	request, err := http.NewRequest(http.MethodPost, fullPath, data)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Set("Content-Type", defaultContentType)
+
+	return request, nil
+}
+
+// CallService call the service provides by plugin server
+func (cli *PluginClient) CallService(service string, in, out interface{}) error {
+	input := new(bytes.Buffer)
+	if in != nil {
+		if err := json.NewEncoder(input).Encode(in); err != nil {
+			return err
+		}
+	}
+
+	body, err := cli.retryCallService(service, input, true)
+	if err != nil {
+		return err
+	}
+	defer body.Close()
+
+	if out != nil {
+		if err := json.NewDecoder(body).Decode(out); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cli *PluginClient) retryCallService(service string, data io.Reader, retry bool) (io.ReadCloser, error) {
+	var start = time.Now()
+	var counter = 0
+
+	for {
+		req, err := cli.newPluginRequest(service, data)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := cli.client.Do(req)
+		if err != nil {
+			if retry {
+				delay := backoff(counter)
+				if timeout(start, delay) {
+					return nil, err
+				}
+				counter++
+				logrus.Warnf("plugin %s call %s failed, retry after %d seconds", cli.address, service, delay)
+				time.Sleep(delay)
+				continue
+			}
+			return nil, err
+		}
+
+		if resp.StatusCode != http.StatusOK {
```

Comment:
Does only 200 make sense?
In my opinion, 2xx and 3xx are both regarded valid.

Reply:
It seems that the `resp` should  be handled out of this function. `101` / `20x` / `30x` are the OK.